### DISCUSSION TITLE
Archiving reddit.com is hell

### DISCUSSION THREAD

**Discussion by ntevenhere** [ArchiveBox/ArchiveBox#1022] [Ideas]
_This post is a sort of stream-of-consciousness exploration. It is both a wishlist, a note to sort my thoughts, and feature ideas. I might even try my hand at writing the necessary changes in the code. Through trying to solve a little itch I have with archiving reddit.com, I explore rigidities I see in archivebox. In advance I thank the maintainers of this project, this post has a lot of wants and not enough thanks. 💚_

The Reddit website is really bloated these days. An archive with all extractors takes 2 minutes and weighs on disk 30MB (each page!). I'm running on a laptop, and the DOM, Singlepage, Readability, basically: anything that relies on Chromium rendering, times out on some reddit.com pages. 
<details>

<summary>log</summary>

```
[+] [2022-09-10 05:33:45] "www.reddit.com/r/test"
    https://www.reddit.com/r/test/
    > ./archive/16xxxxxxxx.xxxxxx
      > title
      > headers
      > singlefile
      > pdf
      > screenshot
      > dom
        Extractor timed out after 60s.
        Run to see full output:
            cd /data/archive/16xxxxxxxx.xxxxxx;
            chromium --headless --no-sandbox --disable-gpu --disable-dev-shm-usage --disable-software-rasterizer --run-all-compositor-stages-before-draw --hide-scrollbars "--user-agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.61 Safari/537.36 ArchiveBox/{VERSION} (+https://github.com/ArchiveBox/ArchiveBox/)" --window-size=1440,2000 --timeout=60000 --dump-dom https://www.reddit.com/r/test/

      > wget
      > readability
      > mercury
        135 files (30.0 MB) in 0:02:12s
```
</details>

A few feature ideas to the end of archiving reddit.com better:

- [ ] Reddit has other, more lightweight mirror domains like https://old.reddit.com and http://i.reddit.com and 3rd parties like https://teddit.net. These have the same content. Perhaps I should be able to configure archivebox to use one of those even when I ask it to archive a www.reddit.com link. This could be done by either:
  - Have some kind of _redirect_, or _alias_ configuration, where I could configure key-pairs that define domain synonyms like `www.reddit.com -> old.reddit.com` and `twitter.com -> nitter.net` and`nitter.* -> nitter.net`. The effect applies doubly to retrieval and archival: `archivebox add www.reddit.com` would archive old.reddit.com, and _searching_ in the web UI would be aware of this "symbolic link".
    - In practice (code,) the user would configure key-pairs (like a dict) or a list of pairs. These values, of course, are ready to be passed to `re.sub()`.
      ```py
      URL_FILTER = [ 
      "^(?:https?://)?(?:www\.)?reddit\.com/(.*)", r"old.reddit.com/\1", 
      "^(?:https?://)?(?:www\.|mobile\.)?twitter.com/(.*)", r"https://nitter.net/\1"]
      ```
     - Or perhaps a user-defined filter function would give the savvy user more freedom and responsibility, and would be one less thing to maintain for archivebox's mantainers unlike the previous point.
  - I guess I could also stop begging for features like an ungrateful fuck, and pre-filter the links *myself* to feed into archivebox like uhhh `sed s/www.reddit.com/old.reddit.com/`. I won't be able to use fancy features like getting archivebox to find urls in my txt files for me `archivebox add < mynotes.txt`, I'd have to filter them beforehand, so `grep <url regex> mynotes.txt | sed <you know> | archivebox add`, not as short, but not a big deal. Though filtering links this way will ultimately be very limiting. Archivebox can parse all of these sources in the `add` command: `auto,pocket_api,wallabag_atom,pocket_html,pinboard_rss,shaarli_rss,medium_rss,netscape_html,rss,json,html,txt,url_list`, but I can't benefit from those parsers if I want to apply my url filters. I would have to parse the source myself, which for some sources isn't as trivial.
    - I propose that archivebox cli could offer parsing and "adding" in separate steps. The `add` command would remain the same, but there would be an additional `parse` command. Example usage: 
      ```sh
      archivebox parse < pocket-export.html
      # Outputs the links found inside pocket-export.html as a list

      # it means I can finally do this:
      archivebox parse < pocket-export.html | sed [whatever] | archivebox add
      ```
- [ ] Or, wanna know what I discovered? With a cookie, Reddit.com *can* serve you a lightweight page that is actually the same as old.reddit.com but you stay on the www. domain. The cookie is: `redesign_optout=true`!
  - Need *one* unified way to set cookies. I did configure my COOKIES_FILE=cookies.txt, it works; Wget does download the lightweight page, but the option is only for wget sadly. All the chromium-based extractors keep saving me rather costing me tens of megabytes of crap. I wish I could set cookies.txt in a unified way. I know that you can already, manually, set up a chromium install and set cookies in a graphical session. But I think Archivebox could certainly engineer to unify COOKIES_FILE, for example, single-file has the option `--browser-cookies-file` and it accepts the same Netscape format as wget. 
    - In the near future I will use `SINGLEFILE_ARGS=["--browser-cookies-file=/data/cookies.txt"]`.
    - mercury-parser has [`--header.name=value`](https://github.com/postlight/parser/pull/337) so in theory I'd be able to set `MERCURY_ARGS=["--header.Cookie=redesign_optout=true"]`
    - Eh... It's more difficult for other extractors that are closer to rawdogging chrome (like pdf, dom and screenshot), but I think custom cookies could still be set for them. There's talk ([#51](https://github.com/ArchiveBox/ArchiveBox/issues/51)) of injecting custom javascript on page loads with puppeteer or selenium, and these two certainly have the ability to set cookies. Vanilla javascript can also set cookies, perhaps before the page loads.

- [ ] Finally, most importantly, after mentioning such specific options, I want the flexibility of toggling certain configurations per url. I want to use custom settings, like SAVE_DOM=False only in reddit.com links.
  - The configuration file would probably have to have a sort of conditional block, where (in pseudo-syntax):
    ```
    if "(www)?\.reddit\.com"
        SAVE_WGET=True
        SAVE_DOM=False
        SAVE_SCREENSHOT=False
        SAVE_PDF=False
        #etc
    ```

(I realize that point 1. and 3. could be satisfied in practice by puppeting archivebox with a hackish script, analyzing urls and setting env variables myself. My script would become the only interface with archivebox to satisfy to my wants, meaning things like the Add interface web UI would be unusable)

The end.

Edit:

The reader may find these features request strange. There are two ways to use archivebox. **(1) one-at-a-time, or (2) batch**. In spite that batch is just the repeated version of one-at-a-time, I feel that it needs special software support. The difference of 1 and 2 might be compared to the needs of REPL interaction and a written program. 
1. A user that enjoys saving snippets as he surfs the web, archives one link at a time and can toggle settings with care for each link's requirements.
2. However, my use is fighting link-rot of external references in my bookmarks and plaintext notes, so I import their entire contents to archivebox periodically. I add thousands of links at a time and I don't want to intervene manually in every link that goes wrong. I wish I could write rules for the edgecase links.

**Comment by pirate** [comment]
I recommend archiving an alternative Reddit Frontend like teddit as you mention. Native Reddit web is indeed hell to archive.

Automatic link rewriting with regex `s/Reddit.com/teddit.com/` is a good idea, I'd definitely support adding that feature. I'm unlikely to have time to add it myself soon but I'd welcome any PRs / implementation discussions on this ticket.

As for per-domain settings, I'd also support adding that. It would increase complexity quite a bit but it seems worth it. For now note that you can change settings for scheduled archiving per entry, but those settings only apply to that schedule entry and not if URLs are added from the add page or CLI manually.

**Reply by TomK32** [reply]
Looks like teddit.com is now a spam website unless you meant teddit.net which was a thing but is now gone.

**Reply by TomLucidor** [reply]
There still might be proxy sides out there but I am not sure anymore
